src/init.c

```diff
@@ -32,11 +32,95 @@
 #include <winsock2.h>
 #endif
 
+# define _CONSTRUCTOR __attribute__((constructor))
+# define _DESTRUCTOR __attribute__((destructor))
+
+#ifdef HAVE_PTHREAD
+
+# include <pthread.h>
+
+#define SSH_STATIC_MUTEX(mutex) \
+    static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER
+
+# define SSH_STATIC_MUTEX_LOCK(mutex) \
+    pthread_mutex_lock(&mutex)
+
+# define SSH_STATIC_MUTEX_UNLOCK(mutex) \
+    pthread_mutex_unlock(&mutex)
+
+#else /* HAVE_PTHREAD */
+
+# define SSH_STATIC_MUTEX(mutex)
+# define SSH_STATIC_MUTEX_LOCK(mutex)
+# define SSH_STATIC_MUTEX_UNLOCK(mutex)
+
+#endif /* HAVE_PTHREAD */
+
+SSH_STATIC_MUTEX(ssh_init_mutex);
+
+/* Counter for initializations */
+static int _ssh_initialized = 0;
+
+/* Cache the returned value */
+static int _ssh_init_ret = 0;
+
+static int _ssh_init(unsigned constructor) {
+
+    int rv = 0;
+
+    if (!constructor) {
+        SSH_STATIC_MUTEX_LOCK(ssh_init_mutex);
+    }
+
+    _ssh_initialized++;
+
+    if (_ssh_initialized > 1) {
+        rv = _ssh_init_ret;
+        goto _ret;
+    }
+
+    if((rv = ssh_threads_init()))
```

Comment:
Always write:

```
rc = ssh_threads_init();
if (rc ...)
```

This makes it easier to debug!

Reply:
Addressed in c4cedf8